### DISCUSSION TITLE
Magic link passwordless authentication

### DISCUSSION THREAD

**Discussion by yordis** [keycloak/keycloak#8962] [Ideas]
### Description

Jira Ticket: https://issues.redhat.com/browse/KEYCLOAK-17578

As an admin user,

I would like to add passwordless authentication to my authentication and registration. Such authentication will send a magic link and/or code to an email and/or SMS that I can use to authenticate

### Reference:

- https://auth0.com/passwordless
- https://www.okta.com/blog/2020/09/magic-links/

**Comment by mabartos** [comment]
@yordis It'd be great to have the "Magic link" authentication method included in Keycloak as it'd be also suitable for 2FA. IMHO, it's a step forward and it can be also useful for adaptive MFA. It's very similar to the Keycloak reset credential approach, where the email is sent to the user with a token, so there might be only a few custom changes in order to accomplish the goal. It can be also very useful with the Trusted devices feature in the future as it could be used as the complete passwordless approach.

@mposolda WDYT?

**Reply by hmlnarik** [reply]
Such authentication is already possible via action tokens.

A custom implementation of action token is in [keycloak quickstarts](https://github.com/keycloak/keycloak-quickstarts/tree/latest/action-token-authenticator). Note that in that quickstart, invocation of external application is shown, but that one is not necessary at all, leading to the "magic link" functionality.

**Reply by mabartos** [reply]
Oh, that's nice. Thanks for that. It'd be good to include it among the other authenticators and extend our possibilities. We'd just need to have more opinions on this.

**Reply by codespearhead** [reply]
[Here's the updated Keycloak Quickstart link](https://github.com/keycloak/keycloak-quickstarts/tree/latest/extension/action-token-authenticator). I haven't tried it myself for passwordless auth though, have you @mabartos ?

**Reply by mabartos** [reply]
@codespearhead Nope, I haven't. I'm not sure what's the process behind promoting stuff from quickstarts to the main codebase, but I'd suggest creating an RFE in this repository. It would be good to express all the requirements, explain the motivation around the promotion, and what value it might bring to Keycloak. Additionally, it'd be great if users would react to the RFE to show it's the thing the community wants.

**Comment by pareshppatelhuf** [comment]
Please check code given in below link. `https://github.com/9228883664/keycloak-otp-password-authenticator`

You can use keycloak for OTP validation as primary login. Many portal want either user can login with username/password or can do login by entering registered mobile number and enter received OTP on portal to login. Valid token needs to be generated by calling keycloak API. This plugin serve the purpose of both user can login with either password or OTP. OOTB keycloak does not provide this feature, instead it provide 2F OTP authenticator with well-known OTP application.

**Comment by FStefanni** [comment]
Hi,

I would like to have an OTP sent by email, as a second auth factor, so if I understand,
this feature request seems exactly what I need.
Any news?

Also, please note that the various links in the answers are currently broken...

Regards

**Comment by MGLL** [comment]
Hi,

Just to let you know that [phasetwo](https://phasetwo.io/docs/introduction/open-source) did an extension for it. You can have a look to the github repo [here](https://github.com/p2-inc/keycloak-magic-link).

If you need more info or details, I think @xgp can give you some.

Best regards,

**Comment by FStefanni** [comment]
Hi,

@MGLL  thank you for the link, I already know that extension, but its license does not permit its use :(
Basically, except use it for yourself, you cannot use it for a commercial purpose.
Here the link to its license:

https://github.com/p2-inc/keycloak-magic-link/blob/main/COPYING

So it seems that we are currently missing a **true** open source plugin for it.

Regards

**Reply by MGLL** [reply]
Hi @xgp, can you provide some info on that (I'm also interested as I might have misunderstood).

Based on 
`You may not provide the software to third parties as a hosted or managed
service, where the service provides users with access to any substantial set of
the features or functionality of the software.`

and
![image](https://github.com/keycloak/keycloak/assets/16251642/953175e6-002d-47fb-a094-81adc21d6dd7)
from the [license change article ](https://phasetwo.io/blog/licensing-change) and this [page](https://phasetwo.io/docs/introduction/open-source).

```
Open source license
With the Elastic License, you can use and host the Phase Two Keycloak extensions as part of your own product. You may not sell  
the extensions as a hosted or managed service, which includes bundling and distribution by companies who sell their products for  
on-prem and private cloud use. You are free to create derivative works, but they must maintain the same license, copyright and  
other notices. You may not represent the Phase Two Keycloak extensions as your own work of authorship.
```

What _I understood_, is that you can use it in your product (SaaS Applications or Platform), but you are **not allowed** to offer it as a managed service to another party (like `okta` could do). Because it's the [phasetwo core product and service offering](https://phasetwo.io/). 
So, if you sell a service based on Keycloak as a managed `Identity and Access Management` to any other company, then you are not allowed to use it.

But I will let @xgp provide a more clear answer to that.

**Reply by xgp** [reply]
> but its license does not permit its use :(
> Basically, except use it for yourself, you cannot use it for a commercial purpose.

@FStefanni This is incorrect. The license permits broad use. You are more than welcome to use it for a commercial purpose.

I think there is some confusion about the Elastic v2 license. The brief history there is that AWS and others starting **selling** hosting of Elasticsearch when it was ASL 2.0. In a move to **keep** Elasticsearch **open source**, but to prohibit cloud providers and resellers from selling hosting of the product itself, they changed their license to what eventually became Elastic v2.

Since then, many companies who are both building open source tools, and extending existing tools, have adopted such a license approach (e.g. CockroachDB, MongoDB, Neo4j, Confluent, etc.). Their licenses restrict cloud providers and other resellers from offering their software **as a service** without purchasing a commercial license. 

Phase Two (the author of the `keycloak-magic-link`extension, and others) found that cloud providers offering Keycloak as a service, were using its extensions, and reselling them as part of their commercial offering. After consideration and a lot of communication with customers and users, we settled on adopting the Elastic v2 license in order to stop such behavior (although we are aware that it continues, now in violation of the license). We attempted to clarify our position on this when we made the change: https://phasetwo.io/blog/licensing-change

That **does not** mean that you can't "use it for a commercial purpose" or "its license does not permit its use". These statements are incorrect and misleading. If you are using it for any use case that does not constitute providing the software itself **as a service**, then that use is permitted.

Some examples:
- If you have created *CatShare*, a paid SaaS service for discovering and sharing cat videos and pictures, and you use Keycloak and `keycloak-magic-links` as your User Management and login mechanism, **this is a permitted use**.
- If you host Keycloak as a service (e.g. Cloud-IAM, Bare.id, Elestio, etc.) and you bundle `keycloak-magic-links` **this is a non-permitted use, and you must obtain a commercial license**.

We are trying to **encourage** the use of our extensions, while **prohibiting** their use by competitors.

**Reply by FStefanni** [reply]
Hi,

@xgp thank you for the clarification. It was not clear to me.
So I'll try to use the extension, since we provide a payed saas, and we want to use keyclock just to manage user logins.

Regards

**Reply by tilmanKranz** [reply]
> * If you have created CatShare, a paid SaaS service for discovering and sharing cat videos and pictures, and you use Keycloak and keycloak-magic-links as your User Management and login mechanism, this is a permitted use.
> * If you host Keycloak as a service (e.g. Cloud-IAM, Bare.id, Elestio, etc.) and you bundle keycloak-magic-links this is a non-permitted use, and you must obtain a commercial license.

The distinction between "paid SaaS service" and "host Keycloak as a service" is not clear to me.

Where can such a license be acquired? I see an offering "Enterprise Keycloak Support Packages" on phasetwo.io, is this the "commercial license"?

**Reply by xgp** [reply]
If you compete with anyone on that list (Phase Two, Cloud-IAM, Bare.id, Elestio, etc.) then you need to contact us (support@phasetwo.io) for a commercial license. There is no information on the website regarding software licenses for our competitors.

**Reply by Wallman** [reply]
@FStefanni maybe you can edit the original question with this answer

**Comment by bahaa** [comment]
Hi,

Using email as 2FA is really a bad idea security wise. It's not really a second factor. If user email got compromised, the attacher can reset the password using the email, and send the OTP/magic link to the email and gain full access to the account.

NIST recommends [against](https://pages.nist.gov/800-63-FAQ/#q-b11) it.

**Comment by codespearhead** [comment]
## Magic Link Passwordless Authentication

It seems the discussion got sidetracked by focusing on "email" and "SMS" for MFA, which serve a different use case than magic links.

Mabartos' suggestion seems to be on the right track, though it hasn't been implemented yet.

If you're interested in this feature, feel free to share your use case here and upvote it by adding a ⬆️ to the discussion’s description.

## SMS and Email for MFA

This question has been asked several times before. You can find previous discussions by searching for “sms” or “email” in the GitHub search bar of the Keycloak repository [3][4].

Currently, there is no official support for using email or phone numbers for authentication in Keycloak. Therefore, any issues you encounter with third-party solutions should be reported in their respective repositories, not here.

If you're interested in this feature, feel free to share your use case here [5], and upvote it by adding a ⬆️ to the discussion’s description after reading this answer carefully [6].

***

[1] https://docs.github.com/en/issues/tracking-your-work-with-issues/administering-issues/closing-an-issue
[2] https://docs.github.com/en/discussions/managing-discussions-for-your-community/managing-discussions#closing-a-discussion
[3] https://github.com/search?q=repo%3Akeycloak%2Fkeycloak+%22SMS%22&type=issues
[4] https://github.com/search?q=repo%3Akeycloak%2Fkeycloak+%22SMS%22&type=discussions
[5] https://github.com/keycloak/keycloak/discussions/27152
[6] https://github.com/keycloak/keycloak/discussions/27152#discussioncomment-10923758

**Comment by Wallman** [comment]
Is there any discussions held about making [p2-inc/keycloak-magic-link](https://github.com/p2-inc/keycloak-magic-link) part of the native Keycloak offering?
